File: go/triangle/triangle.go
package triangle

import (
	"sort"
	"math"
)

type Kind int

const (
    NaT = iota // not a triangle
    Equ = iota // equilateral
    Iso = iota // isosceles
    Sca = iota // scalene
    Deg = iota // degenerate (dig deeper exercise)
)


func KindFromSides(a, b, c float64) Kind {

	sides := []float64{a,b,c}
	sort.Float64s(sides)

	if sides[0] <= 0 || math.IsNaN(sides[0]) || math.IsInf(sides[0], -1) || math.IsInf(sides[2], +1)  { //sides with zero or below zero length or NaN
		return NaT
	} else if sides[0] + sides[1] < sides[2] {
		return NaT
/*
	Dig deeper exercise
	} else if sides[0] + sides[1] == sides[2] {
		return Deg
*/
	} else if sides[0] == sides[1] && sides[0] == sides[2] {
		return Equ
	} else if sides[0] == sides[1] || sides[1] == sides[2] {
		return Iso
	} else {
		return Sca
	}
}

File: haskell/triangle/src/Triangle.hs
module Triangle (TriangleType(..), triangleType) where

import Data.List (sort)

data TriangleType = Equilateral
                  | Isosceles
                  | Scalene
                  | Degenerate
                  | Illegal
                  deriving (Eq, Show)

triangleType :: (Ord a, Eq a, Num a) => a -> a -> a -> TriangleType
triangleType a b c
    | a <= 0 || b <= 0 || c <= 0            = Illegal
    | sides !! 0 + sides !! 1 <  sides !! 2 = Illegal
    | sides !! 0 + sides !! 1 == sides !! 2 = Degenerate
    | a == b && b == c                      = Equilateral
    | a == b || b == c || a == c            = Isosceles
    | otherwise                             = Scalene

    where sides = sort [a, b, c]